crates/accelerate/src/quantum_circuit/circuit_data.rs

```diff
@@ -0,0 +1,642 @@
+#[derive(Clone, Debug)]
+struct PackedInstruction {
+    /// The Python-side operation instance.
+    op: PyObject,
+    /// The index under which the interner has stored `qubits`.
+    qubits_id: IndexType,
+    /// The index under which the interner has stored `clbits`.
+    clbits_id: IndexType,
+}
+
+/// Private wrapper for Python-side Bit instances that implements
+/// [Hash] and [Eq], allowing them to be used in Rust hash-based
+/// sets and maps.
+///
+/// Python's `hash()` is called on the wrapped Bit instance during
+/// construction and returned from Rust's [Hash] trait impl.
+/// The impl of [PartialEq] first compares the native Py pointers
+/// to determine equality. If these are not equal, only then does
+/// it call `repr()` on both sides, which has a significant
+/// performance advantage.
+#[derive(Clone, Debug)]
+struct BitAsKey {
+    /// Python's `hash()` of the wrapped instance.
+    hash: isize,
+    /// The native Py pointer for the instance.
+    id: usize,
+    /// The wrapped instance.
+    bit: PyObject,
+}
+
+impl BitAsKey {
+    fn new(bit: &PyAny) -> PyResult<Self> {
+        Ok(BitAsKey {
+            hash: bit.hash()?,
+            id: bit.as_ptr() as usize,
+            bit: bit.into_py(bit.py()),
+        })
+    }
+}
+
+impl Hash for BitAsKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write_isize(self.hash);
+    }
+}
+
+impl PartialEq for BitAsKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+            || Python::with_gil(|py| {
+                self.bit
+                    .as_ref(py)
+                    .repr()
+                    .unwrap()
+                    .eq(other.bit.as_ref(py).repr().unwrap())
+                    .unwrap()
+            })
+    }
+}
+
+impl Eq for BitAsKey {}
+
+/// A container for :class:`.QuantumCircuit` instruction listings that stores
+/// :class:`.CircuitInstruction` instances in a packed form by interning
+/// their :attr:`~.CircuitInstruction.qubits` and
+/// :attr:`~.CircuitInstruction.clbits` to native vectors of indices.
+///
+/// Before adding a :class:`.CircuitInstruction` to this container, its
+/// :class:`.Qubit` and :class:`.Clbit` instances MUST be registered via the
+/// constructor or via :meth:`.CircuitData.add_qubit` and
+/// :meth:`.CircuitData.add_clbit`. This is because the order in which
+/// bits of the same type are added to the container determines their
+/// associated indices used for storage and retrieval.
+///
+/// As a convenience, :attr:`.CircuitData.qubits` and
+/// :attr:`.CircuitData.clbits` are exposed to Python as ``list`` instances,
+/// which are updated inplace as new bits are added via
+/// :meth:`~.CircuitData.add_qubit` and :meth:`~.CircuitData.add_clbit`.
+/// **DO NOT MODIFY THESE LISTS DIRECTLY**,
+/// or they will become out of sync with the internals of this class.
+/// In this way, a :class:`.CircuitData` owns its ``qubits`` and ``clbits``.
```

Comment:
Turns out that it wasn't necessary for these lists to be updated in place.

I've changed the logic to instead replace the lists each time new bits are added in f2646b4. This way, client code won't be able to depend on the lists being updated.

Technically, this may be a breaking change since previously `QuantumCircuit.{qubits, clbits}` would be updated in place as new bits and registers were added to this circuit. Is this worth documenting in the release notes?

I left `CircuitData.{qubits, clbits}` as attributes rather than methods since the underlying values are cached as bits are added. I'm mildly concerned that changing them to methods might convey to clients that the lists are constructed on the fly and thus expensive to compute, which is not true. I could see an argument where they make more sense as methods to convey that the reference cannot be relied upon, but I've documented this in the docstrings.